diff --git a/glict/BUILD b/glict/BUILD
index fac56eb..ff1db43 100644
--- a/glict/BUILD
+++ b/glict/BUILD
@@ -1,4 +1,6 @@
 # -*-Python-*-
+
+load("@rules_cc//cc:defs.bzl", "cc_binary")
 config_setting(
     name = "darwin",
     values = {"host_cpu": "darwin"}
diff --git a/glict/GLICT/BUILD b/glict/GLICT/BUILD
index e6725f5..bd6b58c 100644
--- a/glict/GLICT/BUILD
+++ b/glict/GLICT/BUILD
@@ -1,5 +1,7 @@
 # -*-Python-*-
 
+load("@rules_cc//cc:defs.bzl", "cc_library")
+
 config_setting(
     name = "windows",
     values = {"cpu": "x64_windows"},
diff --git a/MODULE.bazel b/MODULE.bazel
new file mode 100644
index 0000000..23cce1d
--- /dev/null
+++ b/MODULE.bazel
@@ -0,0 +1,56 @@
+module(
+    name = "glict",
+    version = "0.0.0",
+)
+
+bazel_dep(name = "rules_cc", version = "0.2.14")
+
+bazel_dep(name = "rules_libsdl12", version = "0.0.0")
+
+libsdl12 = use_extension("@rules_libsdl12//:libsdl12.bzl", "libsdl12_extension")
+use_repo(
+    libsdl12,
+    "libsdl12",
+    "x11repository",
+    "xcbrepository",
+    "libx11-dev",
+    "libxext-dev",
+    "libx11-6",
+    "x11proto-dev",
+    "libxrender-dev",
+    "libxcb1-dev",
+    "libxrandr-dev",
+    "libxau-dev",
+    "libx11-xcb-dev",
+    "libxdmcp-dev",
+    "xorg-sgml-doctools",
+    "libxcb1",
+    "libxau6",
+    "libxdmcp6",
+    "libalsaplayer-dev",
+    "libasound2-dev",
+    "libgl-dev",
+    "libglu1-mesa-dev",
+    "libgl1",
+    "libglu1-mesa",
+    "libglvnd0",
+    "libglx-mesa0",
+    "libglx0",
+    "libglapi-mesa",
+    "libx11-xcb1",
+    "libxext6",
+    "libxfixes3",
+    "libxxf86vm1",
+    "libxcb-glx0",
+    "libxcb-dri2-0",
+    "libxcb-dri3-0",
+    "libxcb-present0",
+    "libxcb-sync1",
+    "libxcb-xfixes0",
+    "libxshmfence1",
+    "libdrm2",
+    "libexpat1",
+    "libbsd0",
+    "libmd0",
+    "libxcb-shm0",
+)
